### ISSUE TITLE
read.ctd.odf() is converting IPTS-68 to ITS-90 but not updating the unit

### ISSUE THREAD

**Issue by richardsc** [dankelley/oce#1801]
**ISSUE 1**

Discovered in cioos-siooc/cioos-siooc_data_transform#65, it appears that `read.ctd.odf()` (which I think is what is triggered from a `read.oce()` call) is converting IPTS-68 temperatures to ITS-90 but not updating the unit correctly. Frankly, I think that it *shouldn't* be changing the temperature, but should retain whatever was read from the file.

reprex to come later, but here is an example file (with extension changed to .txt for Github):

[D18667016.txt](https://github.com/dankelley/oce/files/6250866/D18667016.txt)

**Comment by richardsc** [comment]
Also need to check if `as.ctd(read.odf(file))` is doing the right thing, because I see: 

```
> as.ctd(read.odf('D18667016.ODF'))[['units']][['temperature']]
$unit
expression(degree * C)

$scale
[1] "ITS-90"
```

but also
```
> as.ctd(read.odf('D18667016.ODF'))[['temperature']] - dd$V3
  [1] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [38] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 [75] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
[112] 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
```
which suggests that the above is changing the unit but not the actual values ....

**Comment by richardsc** [comment]
Also, the help page for `read.odf()` has the section:

```
Caution:                                                                                                                                                                                                                                             ODF files do not store information on the temperature or salinity
     scale, and ‘read.odf()’ assumes these to be ITS-90 and PSS-78,
     respectively. These scales may be incorrect for old data files.
     Note that the temperature scale can be converted from old scales
     using ‘T90fromT68()’ and ‘T90fromT48()’, although the change will
     be in a fraction of a millidegree, which probably exceeds
     reasonable confidence in old data.
```

which appears to not be generally true (the file above definitely has a temperature scale).

**Comment by dankelley** [comment]
When this gets figured out, it will be very helpful to have tests in the test suite, so that problems cannot arise later.  Conveniently, a test-suite sort of coding is also perfect for making a reprex.

**Comment by dankelley** [comment]
After discussion with @richardsc, I took over this issue.

One item we discussed was our agreement that oce should store the temperatures in the unit of the ODF file, but that `[[` should deliver values on the ITS90 scale, regardless of the original scale.  The reprex below shows that doing so would be consistent with the rest of the package, as exemplified by the built-in `ctd` dataset.

<details><summary>Reprex</summary>

``` r
library(oce)
#> Loading required package: gsw
#> Loading required package: testthat
#> Loading required package: sf
#> Linking to GEOS 3.8.1, GDAL 3.1.4, PROJ 6.3.1
data(ctd)

# Unit and scale of temperature
ctd@metadata$units$temperature
#> $unit
#> expression(degree * C)
#> 
#> $scale
#> [1] "IPTS-68"

# Accessor vs stored values
head(ctd@data$temperature)
#> [1] 14.2245 14.2299 14.2285 14.2256 14.2301 14.2366
head(ctd[["temperature"]])
#> [1] 14.22109 14.22649 14.22509 14.22219 14.22669 14.23318

# Demo that accessor gives T90 value
all.equal(ctd[["temperature"]], T90fromT68(ctd@data$temperature))
#> [1] TRUE
```

<sup>Created on 2021-04-03 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by dankelley** [comment]
I've added some 'ISSUE' labels so I can keep track of the sub-issues.

Actually, I think `read.ctd()`, `read.ctd.odf()` and `read.oce()` are all producing the same results for the temperature unit and stored value, and the results are as we want (IPTS68 for the scale).

**ISSUE 4**

*However*, the `[[` is not working properly, as can be seen in the "should be" parts in the reprex given below.  This tells me the problem: it must be that the `[[` operator for the `ctd` class is checking the scale, but that it is not being called for the `odf` and `oce` classes.

<details>
``` r
# Test read.odf(), read.ctd.odf() and read.oce() with an ODF file supplied with oce
library(oce)
#> Loading required package: gsw
#> Loading required package: testthat
#> Loading required package: sf
#> Linking to GEOS 3.8.1, GDAL 3.1.4, PROJ 6.3.1
file <- system.file("extdata", "CTD_BCD2014666_008_1_DN.ODF.gz", package="oce")

# Read temperatures straight from file
Tref <- read.table(file, skip=675, header=FALSE)$V4
Tref90 <- T90fromT68(Tref)

# read.odf()
ctd1 <- read.odf(file)
#> Warning in read.odf(file): "CRAT_01" should be unitless, but the file states the
#> unit as "S/m" so that is retained in the object metadata. This will likely cause
#> problems. See ?read.odf for an example of rectifying this unit error.
ctd1@metadata$units$temperature
#> $unit
#> expression(degree * C)
#> 
#> $scale
#> [1] "IPTS-68"
all.equal(ctd1@data$temperature, Tref)
#> [1] TRUE
all.equal(ctd1[["temperature"]], Tref90) # should be TRUE
#> [1] "Mean relative difference: 0.0002399424"
all.equal(ctd1[["temperature"]], Tref)   # should be FALSE
#> [1] TRUE

# read.ctd.odf()
ctd2 <- read.ctd.odf(file)
#> Warning in read.odf(file = file, columns = columns, exclude = exclude, debug =
#> debug - : "CRAT_01" should be unitless, but the file states the unit as "S/m" so
#> that is retained in the object metadata. This will likely cause problems. See ?
#> read.odf for an example of rectifying this unit error.
all.equal(ctd2@metadata$units$temperature, ctd1@metadata$units$temperature)
#> [1] TRUE
all.equal(ctd1@data$temperature, ctd2@data$temperature)
#> [1] TRUE
all.equal(ctd2[["temperature"]], Tref90) # should be TRUE
#> [1] TRUE
all.equal(ctd2[["temperature"]], Tref)   # should be FALSE
#> [1] "Mean relative difference: 0.00024"

# read.oce()
ctd3 <- read.oce(file)
#> Warning in read.odf(file, ...): "CRAT_01" should be unitless, but the file
#> states the unit as "S/m" so that is retained in the object metadata. This will
#> likely cause problems. See ?read.odf for an example of rectifying this unit
#> error.
all.equal(ctd3@metadata$units$temperature, ctd1@metadata$units$temperature)
#> [1] TRUE
all.equal(ctd1@data$temperature, ctd3@data$temperature)
#> [1] TRUE
all.equal(ctd3[["temperature"]], Tref90) # should be TRUE
#> [1] "Mean relative difference: 0.0002399424"
all.equal(ctd3[["temperature"]], Tref)   # should be FALSE
#> [1] TRUE
```

<sup>Created on 2021-04-03 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by dankelley** [comment]
I see, from the details (an extract from `R/odf.R`) that the `odf` class does not specialize `[[, so I only need to look at the base class.

<details>

```R
setMethod(f="[[",
          signature(x="odf", i="ANY", j="ANY"),
          definition=function(x, i, j, ...) {
              callNextMethod()         # [[
          })
```

**Comment by richardsc** [comment]
Wow, I had of course completely forgotten about the (documented) behaviour for the `[[` accessor ... That might explain some of the issues as you point out.
```
     It should be noted that the accessor is set up to retrieve
     quantities in conventional units. For example, ‘read.ctd.sbe()’ is
     used on a ‘.cnv’ file that stores pressure in psi, it will be
     stored in the same unit within the ‘ctd’ object, but
     ‘x[["pressure"]]’ will return a value that has been converted to
     decibars.  (To get pressure in PSI, use ‘x[["pressurePSI"]]’.)
     Similarly, temperature is returned in the ITS-90 scale, with a
     conversion having been performed with ‘T90fromT68()’, if the
     object holds temperature in IPTS-68.  Again, temperature on the
     IPTS-68 scale is returned with ‘x@data$temperature’.
```

**Comment by dankelley** [comment]
Re ISSUE 3, I think I should just delete that paragraph from the docs.  I assume it might have been true once (or for some files I looked at a long time ago), but it might also be that I never really looked carefully at the files.

**Comment by dankelley** [comment]
I have everything but `as.ctd()` fixed now, and I have located the problem in `R/ctd.R`, viz.
```R
    ## Default some units (FIXME: this may be a bad idea)
    if ("salinity" %in% dataNames && !("salinity" %in% unitsNames))
        res@metadata$units$salinity <- list(unit=expression(), scale="PSS-78")
    if ("temperature" %in% dataNames && !("temperature" %in% unitsNames))
        res@metadata$units$temperature <- list(unit=expression(degree*C), scale="ITS-90")
    if ("pressure" %in% dataNames && !("pressure" %in% unitsNames))
        res@metadata$units$pressure <- list(unit=expression(dbar), scale="")
 ```
um, yup, it *was* a bad idea!  

As Kate Wolf wrote in The Great DIvide,

> I've been siftin' through the layers 
Of dusty books and faded papers 
They tell a story I used to know 
And it was one that happened so long ago

**Comment by dankelley** [comment]
I think I've addressed all of the issues in "develop" commit df3c7ebc825b10008f5abae89f1ad0b27355b35a.  This involved changes to a few files, and additions to some test files.  Things seem OK for the `read` functions and `[[`, plus for `as.ctd()` in cases of (a) specifying vectors for salinity, temperature and pressure, (b) first arg being an `odf` object, and (c) first arg being an argo object.

I suspect there may be other cases that I have not thought of, but the point is that "develop" now should work on the problem mentioned in this issue, plus in other issues.

To see my tests for the issue at hand, click "details".

<details>

The following tests are at the end of `tests/testthat/test_odf.R`, and I think they cover all of the problems mentioned in this issue.

```R
test_that("ODF temperature scale, IPTS68 and ITS90", {
          file <- system.file("extdata", "CTD_BCD2014666_008_1_DN.ODF.gz", package="oce")
          # Get IPTS68 temperature, ITS90, and unit, all needed for tests.
          # Note that all read() calls need to check for a warning that results
          # from an error in this ODF file (and quite a few ODF files, it seems).
          Tref68 <- read.table(file, skip=675, header=FALSE)$V4
          Tref90 <- T90fromT68(Tref68)
          unit <- list(unit=expression(degree*C), scale="IPTS-68")
          # read.odf()
          ctd1 <- expect_warning(read.odf(file), "should be unitless")
          expect_equal(ctd1@metadata$units$temperature, unit)
          expect_equal(ctd1@data$temperature, Tref68)
          expect_equal(ctd1[["temperature"]], Tref90)
          # read.ctd.odf()
          ctd2 <- expect_warning(read.ctd.odf(file), "should be unitless")
          expect_equal(ctd2@metadata$units$temperature, unit)
          expect_equal(ctd2@data$temperature, Tref68)
          expect_equal(ctd2[["temperature"]], Tref90)
          # read.oce()
          ctd3 <- expect_warning(read.oce(file), "should be unitless")
          expect_equal(ctd3@metadata$units$temperature, unit)
          expect_equal(ctd3@data$temperature, Tref68)
          expect_equal(ctd3[["temperature"]], Tref90)
          # as.ctd(read.odf())
          ctd1B <- as.ctd(ctd1)
          expect_equal(ctd1B@metadata$units$temperature, unit)
          expect_equal(ctd1B@data$temperature, Tref68)
          expect_equal(ctd1B[["temperature"]], Tref90)
          # as.ctd(read.ctd.odf())
          ctd2B <- as.ctd(ctd2)
          expect_equal(ctd2B@metadata$units$temperature, unit)
          expect_equal(ctd2B@data$temperature, Tref68)
          expect_equal(ctd2B[["temperature"]], Tref90)
          # as.ctd(read.oce())
          ctd3B <- as.ctd(ctd3)
          expect_equal(ctd3B@metadata$units$temperature, unit)
          expect_equal(ctd3B@data$temperature, Tref68)
          expect_equal(ctd3B[["temperature"]], Tref90)
})
```

**Comment by richardsc** [comment]
Looks good! Sorry for taking a few days to check/test.
